src/lua/datetime.lua

```diff
@@ -0,0 +1,605 @@
+    return math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY)
+end
+
+-- convert UTC seconds to local seconds, adjusting by timezone
+local function local_secs(obj)
+    return obj.epoch + obj.tzoffset * 60
+end
+
+local function utc_secs(epoch, tzoffset)
+    return epoch - tzoffset * 60
+end
+
+-- get epoch seconds, shift to the local timezone
+-- adjust from 1970-related to 0000-related time
+-- then return dt in those coordinates (number of days
+-- since Rata Die date)
+local function local_dt(obj)
+    return builtin.tnt_dt_from_rdn(local_rd(local_secs(obj)))
+end
+
+local function datetime_cmp(lhs, rhs)
+    if not is_datetime(lhs) or not is_datetime(rhs) then
+        return nil
+    end
+    local sdiff = lhs.epoch - rhs.epoch
+    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
+end
+
+local function datetime_eq(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == 0
+end
+
+local function datetime_lt(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc < 0
+end
+
+local function datetime_le(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc <= 0
+end
+
+local function datetime_serialize(self)
+    check_date(self, 'datetime.__serialize()')
+    return { epoch = self.epoch, nsec = self.nsec,
+             tzoffset = self.tzoffset, tzindex = 0 }
+end
+
+--[[
+    parse_tzoffset accepts time-zone strings in both basic
+    and extended iso-8601 formats.
+
+    Basic    Extended
+    Z        N/A
+    +hh      N/A
+    -hh      N/A
+    +hhmm    +hh:mm
+    -hhmm    -hh:mm
+
+    Returns timezone offset in minutes if string was accepted
+    by parser, otherwise raise an error.
+]]
+local function parse_tzoffset(str)
+    local offset = ffi.new('int[1]')
+    local len = builtin.tnt_dt_parse_iso_zone_lenient(str, #str, offset)
+    if len ~= #str then
+        error(('invalid time-zone format %s'):format(str), 3)
+    end
+    return offset[0]
+end
+
+local function datetime_new_raw(epoch, nsec, tzoffset)
+    local dt_obj = ffi.new(datetime_t)
+    dt_obj.epoch = epoch
+    dt_obj.nsec = nsec
+    dt_obj.tzoffset = tzoffset
+    dt_obj.tzindex = 0
+    return dt_obj
+end
+
+local function datetime_new_dt(dt, secs, nanosecs, offset)
+    local epoch = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY
+    return datetime_new_raw(epoch + secs - offset * 60, nanosecs, offset)
+end
+
+local function get_timezone(offset)
+    if type(offset) == 'number' then
+        return offset
+    elseif type(offset) == 'string' then
+        return parse_tzoffset(offset)
+    end
+end
+
+local function bool2int(b)
+    return b and 1 or 0
+end
+
+-- create datetime given attribute values from obj
+local function datetime_new(obj)
+    if obj == nil then
+        return datetime_new_raw(0, 0, 0)
+    end
+    check_table(obj, 'datetime.new()')
```

Comment:
I can't explain, but this is a way we need to put to have correct error level. For `check_table` and `check_str` we had to use 2, but for `check_range` - 3. Despite the fact that they are seemingly on the same call level. 🤷‍♀️

I'd love to see any reasonable explanation here ?! @igormunkin any hints why it might be so?

Reply:
This is how Lua TCO works: consider `return` prior to `error` call in `check_table` and `check_str` and none in `check_range`. Hence, `error` call in the `range` case is "deeper" than in the others and the `level` value differs.

For more info see [PIL 6.3](https://www.lua.org/pil/6.3.html).